### PR TITLE
add andThen for validation chaining

### PR DESCRIPTION
**Pull request by colin-lamed**

### CODE REVIEW

**Comment by Profpatsch** [comment]
In which way does it break the laws? I think that should be documented.

**Comment by Profpatsch** [comment]
Ah, so it’s for when you want to keep continuing validation, but have a point where you say “error out now when there’s been any errors so far; don’t try to collect more errors”.

I can see how this is interesting when you have e.g. a parallel pooled computation want to collect all errors for one pool, but cancel as soon as one pool has any errors. But normally you’d use a sliding window for that (up to `k` connections at the same time), so maybe a bad example.

Do you have an example of where you needed this in practice?

**Comment by colin-passiv** [comment]
Ideally, as many errors will be reported as possible, but some validation can be dependent on previous validation. For example

```purescript
validateDate :: String -> V (Array String) DateTime

validateRange :: Tuple DateTime DateTime -> V (Array String) (Tuple DateTime DateTime)
validateRange = (Tuple <$> validateDate d1 <*> validateDate d2) `andThen` validateRange
```

File: src/Data/Validation/Semigroup.purs [hunk 1]
```diff
@@ -56,6 +56,16 @@ isValid _ = false
 toEither :: forall err result. V err result -> Either err result
 toEither (V e) = e
 
+-- | Apply a function if successful, to enable chaining of validation.
+-- |
+-- | Similar to a monadic bind, except it is inconsistent with Apply - that is,
+-- | where as `apply (V err a) (V err a)` accumulates failures,
+-- | `(V err a) ``andThen`` (\a -> V err a)` has fail-fast semantics
```
Comment (LiamGoodacre) [inline src/Data/Validation/Semigroup.purs]: There's a strange mixture of types and terms going on in these examples?
How about:
```purescript
apply (invalid x) (invalid y) = invalid (x <> y)

andThen (invalid x) (\_ -> invalid y) = invalid x
andThen (pure v) b = b v
```

File: src/Data/Validation/Semiring.purs [hunk 2]
```diff
@@ -57,6 +57,16 @@ isValid _ = false
 toEither :: forall err result. V err result -> Either err result
 toEither (V e) = e
 
+-- | Apply a function if successful, to enable chaining of validation.
+-- |
+-- | Similar to a monadic bind, except it is inconsistent with Apply - that is,
+-- | where as `apply (V err a) (V err a)` accumulates failures,
+-- | `(V err a) ``andThen`` (\a -> V err a)` has fail-fast semantics
+-- | (`>>=` would be expected to be consistent).
+andThen :: forall err a b. V err a -> (a -> V err b) -> V err b
```
Comment (LiamGoodacre) [inline src/Data/Validation/Semiring.purs]: Both of these need to be exported.

**Pull request merged**